promql/engine.go

```diff
@@ -668,16 +691,22 @@ func (enh *EvalNodeHelper) signatureFunc(on bool, names ...string) func(labels.L
 
 // rangeEval evaluates the given expressions, and then for each step calls
 // the given function with the values computed for each expression at that
-// step.  The return value is the combination into time series of  of all the
+// step.  The return value is the combination into time series of all the
 // function call results.
 func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ...Expr) Matrix {
 	numSteps := int((ev.endTimestamp-ev.startTimestamp)/ev.interval) + 1
 	matrixes := make([]Matrix, len(exprs))
 	origMatrixes := make([]Matrix, len(exprs))
+	tempNumSamples := 0
 	for i, e := range exprs {
 		// Functions will take string arguments from the expressions, not the values.
 		if e != nil && e.Type() != ValueTypeString {
+			tempNumSamples = ev.currentSamples
 			matrixes[i] = ev.eval(e).(Matrix)
+			// Once inner calls to eval are done we don't need to account for memory for
+			// samples within those calls. However the result of the call to eval is in mem.
+			ev.currentSamples = tempNumSamples
+			ev.currentSamples += len(matrixes[i])
```

Comment:
This is the number of series, not the number of samples. I think you're doing the saving of tempNumSamples in the wrong place, and you don't need any code changes to this part as ev.currentSamples will already have the right value.